PR: https://github.com/odoo/odoo/pull/229665

From: 05e22346050d61adec77f1821b4e901b479c4386
From: SaraBriki
Date: 2026-01-06 08:23:55

Breaking data model changes scores: del:3 + add:0, change matches:
-    user_id = fields.Many2one(
-    allowed_user_ids = fields.Many2many('res.users', compute='_compute_allowed_user_ids', readonly=True)

Total Changes: 41

[IMP] hr*,hr_recruitment**: show recruiters with adequate access rights

*: ,_skills
**: ,_skills,_survey

Issue:
In Recruitment -> Job Position form, the Recruiter dropdown lists all company
users, including those without recruiter rights.

Fix:
* Replace the `user_id` field in `hr.applicant`, `hr_recruitment_reports`, and
  `hr.job` with `recruiter` for clarity, using `hr.employee` instead of
  `res.users`.
* Restrict the dropdown to employees whose linked users have the correct
  recruiter access rights (Recruitment Officer or Administrator).
* Update SQL queries accordingly.

task-5083125

closes odoo/odoo#229665

Related: odoo/enterprise#96434
Related: odoo/upgrade#8565
Signed-off-by: Yannick Tivisse (yti) <yti@odoo.com>

================================= pseudo patch: =================================

--- a/addons/hr/models/hr_job.py
+++ b/addons/hr/models/hr_job.py
@@ -1,7 +1,7 @@
 # -*- coding: utf-8 -*-
 # Part of Odoo. See LICENSE file for full copyright and licensing details.
 
-from odoo import api, fields, models, _
+from odoo import api, fields, models
 from odoo.addons.html_editor.tools import handle_history_divergence
 
 
@@ -11,6 +11,12 @@ class HrJob(models.Model):
     _inherit = ['mail.thread']
     _order = 'sequence'
 
+    def _recruiter_domain(self):
+        return [
+            ("user_id", "!=", False),
+            ("user_id.share", "=", False),
+        ]
+
     active = fields.Boolean(default=True)
     name = fields.Char(string='Job Position', required=True, index='trigram', translate=True)
     sequence = fields.Integer(default=10)
@@ -23,18 +29,17 @@ class HrJob(models.Model):
     employee_ids = fields.One2many('hr.employee', 'job_id', string='Employees', groups='base.group_user')
     description = fields.Html(string='Job Description', sanitize_attributes=False)
     requirements = fields.Text('Requirements', groups="hr.group_hr_user")
-    user_id = fields.Many2one(
-        "res.users",
+    recruiter_id = fields.Many2one(
+        'hr.employee',
         "Recruiter",
-        domain="[('share', '=', False), ('company_ids', '=?', company_id)]",
-        default=lambda self: self.env.user,
+        domain=_recruiter_domain,
+        check_company=True,
+        default=lambda self: self.env.user.employee_id,
         groups="hr.group_hr_user",
         tracking=True,
         help="The Recruiter will be the default value for all Applicants in this job \
             position. The Recruiter is automatically added to all meetings with the Applicant.",
     )
-    # TODO (master): remove the field `allowed_user_ids`.
-    allowed_user_ids = fields.Many2many('res.users', compute='_compute_allowed_user_ids', readonly=True)
     department_id = fields.Many2one('hr.department', string='Department', check_company=True, tracking=True, index='btree_not_null')
     company_id = fields.Many2one('res.company', string='Company', default=lambda self: self.env.company, tracking=True)
     contract_type_id = fields.Many2one('hr.contract.type', string='Contract Type', tracking=True)
@@ -56,28 +61,6 @@ class HrJob(models.Model):
             job.no_of_employee = result.get(job.id, 0)
             job.expected_employees = result.get(job.id, 0) + job.no_of_recruitment
 
-    @api.depends("company_id")
-    def _compute_allowed_user_ids(self):
-        company_ids = self.mapped("company_id.id")
-        domain = [("share", "=", False)]
-        if company_ids:
-            domain += [("company_ids", "in", company_ids)]
-
-        users_by_company = dict(
-            self.env["res.users"]._read_group(
-                domain=domain,
-                groupby=["company_id"],
-                aggregates=["id:recordset"],
-            ),
-        )
-
-        all_users = self.env["res.users"]
-        for users in users_by_company.values():
-            all_users |= users
-
-        for job in self:
-            job.allowed_user_ids = users_by_company.get(job.company_id, all_users)
-
     @api.model_create_multi
     def create(self, vals_list):
         """ We don't want the current user to be follower of all created job """
